### PR TITLE
Better parsing of host/port

### PR DESCRIPTION
**Pull request by cyisfor**
One nitpick I have with vibe is that it can't parse IPv6 addresses in urls (because of colons between []). The parsing of that stuff was kind of clunky too, using nonspecific array splits and trying to examine those to see if there's a good host/port. So, I rewrote the host/port parser. Now it does IPv6 addresses just fine. Verified it with an app of my own. I haven't tested against a huge comprehensive list of all sorts of possible URLs, but that'd be a task in of itself I think. 

IPv6 addresses can have variable amounts of colons, and numbers between colons are often optional too, which is why those difficult to parse addresses are surrounded by [] in URLs, so all I had to do was search for ].

tl;dr with this patch, if you have an IPv6 address, then http://[1:2::2::33]:45/ should work now.

### CODE REVIEW

**Comment by s-ludwig** [comment]
Basically looks good, but I'd like to make this a bit more compact - the `import` and the `struct HostPort` definition could go into `parseHostPort` to make it more self-contained. You are using a manual state machine, is there a reason why using `find` or `indexOf` wouldn't work? It would squash everything together considerably:
```
if (s.startsWith('[')) {
    auto idx = s.indexOf(']');
    enforce(idx > 0, "Missing closing ']' for IPv6 address.");
    ret.host = s[1 .. idx];
    s = s[idx+1 .. $];
} else {
    auto idx = s.indexOf(':');
    if (idx < 0) idx = s.length;
    enforce(idx > 0, "Missing host name.");
    ret.host = s[0 .. idx];
    s = s[idx .. $];
}
if (s.startsWith(':')) ret.port = s[1 .. $].to!ushort;
```

Corresponding issue: #1388
